### DISCUSSION TITLE
How to Obtain Detailed Traces for Chatbot Created with Ragflow?

### DISCUSSION THREAD

**Discussion by erwan-simon-CATIE** [infiniflow/ragflow#2610] [Q&A] [answered]
Hello everyone,

I’m currently exploring Ragflow and I’m interested in capturing detailed traces of the chatbot's process at each stage. The goal is to thoroughly analyze how the system behaves.

The kind of stuff that interest me are:

- The prompt passed to the LLM at the final stage (and idealy, at each steep of the pipeline)
- How the context is modified by the reranker (ei. have chunk list before and after for comparaison)
- ... Any other intermediary steps that contribute to the final output ! :-)

To achieve this, I was considering the possibility of connecting Ragflow to a Langfuse instance, which seems like a promising solution for trace logging and monitoring. However, I'm unsure about the best way to proceed.

Here are a few specific questions I have:

- Is there any built-in functionality within Ragflow that allows capturing such detailed process traces?
- Has anyone attempted to integrate Ragflow with Langfuse (or a similar tracing tool)? If so, how can this be done effectively?
- Are there any recommended best practices or tools to track the flow, especially for analyzing how prompts evolve and how decisions are made at each stage?
- Any insights or suggestions on this would be highly appreciated. Looking forward to your feedback!

Thank you in advance.

Best regards,

**Comment by erwan-simon-CATIE** [comment, marked as answer]
For anyone looking for a way to access the query trace (for testing purposes)... just click on the light bulb icon :-)
![2024-09-26_14-14](https://github.com/user-attachments/assets/02ef13fc-e8e9-4759-a454-dc821d9a6d48)

**Reply by zlkang** [reply]
There is no light bulb icon.
<img width="117" alt="light" src="https://github.com/user-attachments/assets/91603d13-a086-43d8-b7e0-b627dad25aec" />

**Reply by KevinHuSh** [reply]
Do chat again.

**Comment by KevinHuSh** [comment]
The 'retrival test' in knowledgebase can also help you to understand the data before sending to llm.

**Comment by Sos-Zachary** [comment]
@erwan-simon-CATIE hi there! I'm very glad that someone shares the same idea with me. I'm currently using the API provided by RAGflow for other users to access. It's necessary to keep track of each user's Q&A, the matching degree of document retrieval, and receive user evaluations. Langfuse has achieved these key points. Now I'm also confused about what the best practice is for building Langfuse? Do you have any new insights recently? I would be very grateful if you could share them!

**Reply by KevinHuSh** [reply]
Langfuse is a good feature.
